### PR TITLE
Fastsim fix ele track ref

### PR DESCRIPTION
**Pull request by lveldere**
This fixes an issue that caused gsfElectrons with tracker-driven seeds not to be considered as electrons by the particle flow, because a broken link between gsfElectron and recoTrack. 

In FastSim, Tracker-driven electron seeds are produced from the track collection before pile-up mixing,
because the algorithm depends on trajectories which are only available before mixing.
Because of this, tracker-driven electrons are linked to the track collection before pile-up mixing.
Particle flow only considers the after-mixing track collection and thus gets confused.

To fix this, A new module, ElectronSeedTrackRefFix, is implemented and inserted in the electron reconstruction sequence. This modules makes a copy of the tracker-driven electron seeds while replacing in each seed,
the reference to the "before mixing" track with a reference to the corresponding "after mixing" track.

### CODE REVIEW

**Comment by lveldere** [comment]
please test

**Comment by lveldere** [comment]
+1

File: FastSimulation/Configuration/python/FamosSequences_cff.py [hunk 1]
```diff
@@ -203,7 +203,24 @@
...
+    "ElectronSeedTrackRefFix",
+    PreGsfLabel = trackerDrivenElectronSeedsTmp.PreGsfLabel,
+    PreIdLabel = trackerDrivenElectronSeedsTmp.PreIdLabel,
+    oldTrackCollection = trackerDrivenElectronSeedsTmp.TkColList[0],
+    newTrackCollection = cms.InputTag("generalTracks"),
+    seedCollection = cms.InputTag("trackerDrivenElectronSeedsTmp",trackerDrivenElectronSeedsTmp.PreGsfLabel.value()),
+    idCollection = cms.InputTag("trackerDrivenElectronSeedsTmp",trackerDrivenElectronSeedsTmp.PreIdLabel.value())
+    )
+print trackerDrivenElectronSeeds.seedCollection
+print trackerDrivenElectronSeeds.idCollection
```
Comment (davidlange6) [inline FastSimulation/Configuration/python/FamosSequences_cff.py]: remove these two printouts

File: FastSimulation/Configuration/python/FamosSequences_cff.py [hunk 2]
```diff
@@ -203,7 +203,24 @@
 electronGsfTracks.TTRHBuilder = 'WithoutRefit'
 electronGsfTracks.TrajectoryInEvent = True
 
+from RecoParticleFlow.PFTracking.trackerDrivenElectronSeeds_cff import *
+from RecoParticleFlow.PFTracking.mergedElectronSeeds_cfi import *
+
+trackerDrivenElectronSeedsTmp = trackerDrivenElectronSeeds
```
Comment (davidlange6) [inline FastSimulation/Configuration/python/FamosSequences_cff.py]: this should be a clone, right?
Eg  = trackerDrivenElectronSeeds.clone()

**Comment by lveldere** [comment]
Implemented comments from @davidlange6
I don't think doing the clone has an impact here, but I guess it's safer.

BTW: is the effect of a clone wrt a simple = documented anywhere?

**Comment by davidlange6** [comment]
> On Mar 9, 2015, at 5:03 PM, lveldere <notifications@github.com> wrote:
> 
> Implemented comments from @davidlange6
> I don't think doing the clone has an impact here, but I guess it's safer.
> 
> BTW: is the effect of a clone wrt a simple = documented anywhere?

it doesn’t change the original object.

**Comment by lveldere** [comment]
+1

File: FastSimulation/Configuration/python/FamosSequences_cff.py [hunk 3]
```diff
@@ -203,7 +203,21 @@
...
+trackerDrivenElectronSeedsTmp.TkColList = cms.VInputTag(cms.InputTag("generalTracksBeforeMixing"))
+trackerDrivenElectronSeeds = cms.EDProducer(
+    "ElectronSeedTrackRefFix",
+    PreGsfLabel = trackerDrivenElectronSeedsTmp.PreGsfLabel,
+    PreIdLabel = trackerDrivenElectronSeedsTmp.PreIdLabel,
+    oldTrackCollection = trackerDrivenElectronSeedsTmp.TkColList[0],
+    newTrackCollection = cms.InputTag("generalTracks"),
+    seedCollection = cms.InputTag("trackerDrivenElectronSeedsTmp",trackerDrivenElectronSeedsTmp.PreGsfLabel.value()),
+    idCollection = cms.InputTag("trackerDrivenElectronSeedsTmp",trackerDrivenElectronSeedsTmp.PreIdLabel.value())
+    )
```
Comment (davidlange6) [inline FastSimulation/Configuration/python/FamosSequences_cff.py]: @lveldere - sorry, I had missed the redefinition of trackerDrivenElectronSeeds here before. Is it needed or is it just the same as in the trackerDrivenElectronSeeds_cff?

**Comment by lveldere** [comment]
@davidlange6 

Hi David
trackerDrivenElectronSeedsTmp is the same as trackerDrivenElectronSeeds in  trackerDrivenElectronSeeds_cff, except for a different track collection.

**Comment by davidlange6** [comment]
Right, but what about the redefinition of trackerDrivenElectronSeeds just below the clone?

**Comment by lveldere** [comment]
From the description:

"""
A new module, ElectronSeedTrackRefFix, is implemented and inserted in the electron reconstruction sequence. This modules makes a copy of the tracker-driven electron seeds while replacing in each seed,
the reference to the "before mixing" track with a reference to the corresponding "after mixing" track.
"""

With this implemented, PF recognises tracker-driven electrons as electrons.
The copy gets the usual label of the tracker-driven electron seeds,
so that I don't have to make changes further down in the sequences.

**Comment by lveldere** [comment]
BTW, this is still implemented a bit dirty.

I make the assumption the signal tracks are first in the mixed track collection, which is right.
It would be cleaner to make the track mixer spit out a map between the signal tracks before mixing and the mixed track collection. Then I don't have to make such assumptions. That will be something for the next release.

**Comment by davidlange6** [comment]
@lveldere - i see now - this is a bit dangerous as we find out from time to time when two previously separate cffs need to be combined. but here it should be ok assuming no consumers of the original producer in fastsim..

>With this implemented, PF recognises tracker-driven electrons as electrons.
>The copy gets the usual label of the tracker-driven electron seeds,
>so that I don't have to make changes further down in the sequences.

**Comment by davidlange6** [comment]
+1

**Pull request merged**